File: src/crypto/signer.py
"""
Signer: wraps the state and workflow of a single MuSig2-H signing participant.

Thin wrapper that internally calls stateless functions from musig2h.py,
managing keys, nonces, and protocol phases to prevent misuse and nonce reuse.
"""

import random

from src.crypto.musig2h import keygen, presign, sign


class Signer:
    """
    A single MuSig2-H signing participant.

    Lifecycle:
        construct (KeyGen) → set_peers → presign → receive_agg_nonce → sign
        To sign again, must redo presign → receive_agg_nonce → sign
    """

    def __init__(self, seed=None):
        """
        Create a signer, automatically executes KeyGen.

        seed: random seed (int or None), for reproducible testing.
        """
        self.rng = random.Random(seed)
        self.sk, self.pk = keygen(self.rng)

        self._peers = None   # list of other signers' public keys
        self._pp = None      # own nonce commitments (public)
        self._st = None      # own nonce secrets (private)
        self._app = None     # aggregated nonce commitments

    def set_peers(self, peer_pks):
        """
        Set the list of other signers' public keys (excluding self).

        peer_pks: list of curve points
        """
        self._peers = list(peer_pks)

    def presign(self):
        """
        Offline pre-signing: generate 4 nonce groups, return public commitments.

        Returns: pp = (R_1, R_2, R_3, R_4), tuple of curve points
        """
        self._pp, self._st = presign(self.rng)
        self._app = None  # new nonce round, old aggregated commitments invalidated
        return self._pp

    def receive_agg_nonce(self, app):
        """
        Receive aggregated nonce commitments.

        app: (R_1, R_2, R_3, R_4), produced by preagg()
        """
        if self._st is None:
            raise RuntimeError("must call presign() first")
        self._app = app

    def sign(self, message):
        """
        Online signing: compute partial signature after receiving the message.

        message: bytes, the message to sign
        Returns: (R, (s0, s1))

        Nonce secrets are automatically destroyed after signing; must presign again to sign again.
        """
        if self._peers is None:
            raise RuntimeError("must call set_peers() first")
        if self._st is None:
            raise RuntimeError("must call presign() first")
        if self._app is None:
            raise RuntimeError("must call receive_agg_nonce() first")

        out = sign(self._st, self._app, self.sk, self.pk, message, self._peers)

        # destroy nonce after use, prevent reuse which could leak secret key
        self._st = None
        self._pp = None
        self._app = None

        return out

File: src/cpp_musig/signer.cpp
/**
 * Signer state machine — C++ translation of src/crypto/signer.py.
 *
 * Enforces protocol lifecycle and destroys nonce secrets after signing
 * to prevent nonce reuse (which would leak the secret key).
 */

#include "signer.hpp"

Signer::Signer(uint64_t seed) : rng_(seed) {
    auto kp = musig_keygen(rng_);
    sk_ = kp.sk;
    pk_ = kp.pk;
}

void Signer::set_peers(const std::vector<Point>& peer_pks) {
    peers_ = peer_pks;
    peers_set_ = true;
}

void Signer::set_agg_key(const Point& apk, const Scalar& a) {
    cached_apk_ = apk;
    cached_a_ = a;
}

std::array<Point, NU> Signer::presign() {
    auto ps = musig_presign(rng_);
    auto pp = ps.pp;
    presign_state_ = std::move(ps);
    app_ = std::nullopt;  // new nonce round, invalidate old aggregated commitments
    return pp;
}

void Signer::receive_agg_nonce(const std::array<Point, NU>& app) {
    if (!presign_state_) {
        throw std::runtime_error("must call presign() first");
    }
    app_ = app;
}

SignOutput Signer::sign(const std::vector<uint8_t>& message) {
    if (!peers_set_) {
        throw std::runtime_error("must call set_peers() first");
    }
    if (!presign_state_) {
        throw std::runtime_error("must call presign() first");
    }
    if (!app_) {
        throw std::runtime_error("must call receive_agg_nonce() first");
    }

    const Point* apk_ptr = cached_apk_ ? &(*cached_apk_) : nullptr;
    const Scalar* a_ptr = cached_a_ ? &(*cached_a_) : nullptr;

    SignOutput out = musig_sign(
        presign_state_->st,
        *app_,
        sk_, pk_,
        message,
        peers_,
        apk_ptr, a_ptr
    );

    // Destroy nonce after use — prevent reuse
    presign_state_ = std::nullopt;
    app_ = std::nullopt;

    return out;
}
